lib/ember-app.js

```diff
@@ -54,30 +55,49 @@ function EmberApp(options) {
   this._app = AppFactory['default']();
 }
 
-EmberApp.prototype.boot = function() {
-  return this._app.boot();
+EmberApp.prototype.boot = function(options) {
+  return this._app.boot().then(function(app) {
+    return app.buildInstance().boot(options);
+  });
 };
 
-EmberApp.prototype.visit = function(url) {
+EmberApp.prototype.visit = function(request, response) {
+  var path = request.path;
   var doc = new SimpleDOM.Document();
   var rootElement = doc.body;
   var options = { isBrowser: false, document: doc, rootElement: rootElement };
 
-  return this._app.visit(url, options).then(function(instance) {
-    var head;
-    if (doc.head) {
-      head = HTMLSerializer.serialize(doc.head);
-    }
-    try {
-      return {
-        url: instance.getURL(), // TODO: use this to determine whether to 200 or redirect
-        title: doc.title,
-        head: head,
-        body: HTMLSerializer.serialize(rootElement) // This matches the current code; but we probably want `serializeChildren` here
-      };
-    } finally {
-      instance.destroy();
+  return this.boot(options).then(function(instance) {
+    var fastbootService = instance.lookup('service:fastboot');
+
+    var cookies = request.get('cookie');
+    if (cookies) {
+      cookies = cookie.parse(cookies);
     }
+
+    fastbootService.setProperties({
```

Comment:
This still has the problem that cookies are only accessible after the application instance has booted (e.g. you cannot use cookies in initializers/instance initializers). Not sure whether using cookies in (instance) initializer should be considered an anti pattern but I guess it might be surprising to people that they can only access cookies after all instance initializers have run?